### PR TITLE
Add Latest Slasher KV Methods to DB Subpackage

### PR DESCRIPTION
**Pull request by rauljordan**
Part of #8331, this PR adds some latest key-value store methods, especially regarding pruning and fetching of slasher double votes, into the `develop` branch from current code in `feature/slasher`. This is part of an effort to reduce the final diff size of `feature/slasher` in preparation for Prysm v2.

These changes **DO NOT AFFECT** runtime, as the new slasher is not yet live.

The changes include:
- A much simpler way of pruning attestations and proposals now that the schema is simpler
- Efficient way of checking for proposer double votes using a better schema and concurrency

### CODE REVIEW

File: beacon-chain/db/slasherkv/pruning.go [hunk 1]
```diff
@@ -11,199 +11,153 @@ import (
...
-	if currentEpoch < historyLength {
-		log.Debugf("Current epoch %d < history length %d, nothing to prune", currentEpoch, historyLength)
-		return nil
+// PruneProposalsAtEpoch deletes all proposals from the slasher DB with epoch
+// less than or equal to the specified epoch.
+func (s *Store) PruneProposalsAtEpoch(
+	ctx context.Context, maxEpoch types.Epoch,
+) (numPruned uint, err error) {
+	var endPruneSlot types.Slot
+	endPruneSlot, err = helpers.StartSlot(maxEpoch)
```
Comment (terencechain) [inline beacon-chain/db/slasherkv/pruning.go:91]: I think this needs to be `helpers.StartSlot(maxEpoch) - 1`

Comment (rauljordan) [reply beacon-chain/db/slasherkv/pruning.go:91]: @terencechain maybe the func name can be improved, but it relies on pruning up to and including the max epoch specified to it

```
// PruneProposalsAtEpoch deletes all proposals from the slasher DB with epoch
// less than or equal to the specified epoch.
```

File: beacon-chain/db/slasherkv/pruning.go [hunk 2]
```diff
@@ -11,199 +11,153 @@ import (
...
-		proposalBkt := tx.Bucket(proposalRecordsBucket)
-		c := proposalBkt.Cursor()
+		bkt := tx.Bucket(attestationDataRootsBucket)
+		c := bkt.Cursor()
 		k, _ := c.First()
-		lowestSlot = slotFromProposalKey(k)
+		if k == nil {
+			return nil
+		}
+		hasData = true
```
Comment (terencechain) [inline beacon-chain/db/slasherkv/pruning.go:32]: Wonder whether makes sense to add here. Prob not..
```go
if uint64PrefixGreaterThan(k, encodedEndPruneEpoch) {
				return nil
			}
```

**Review by terencechain — APPROVED** [review]
I dont have the complete knowledge to review `beacon-chain/db/slasherkv/slasher.go`, feel free to ask another to review that or merge as it is

**Pull request merged**